src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+                .checksum = checksum,
+                .progress = .{ .table_index = .{ .table = table } },
+            });
+        }
+    }
+
+    pub fn repair_ready(queue: *const GridRepairQueue, address: u64, checksum: u128) bool {
+        const fault_index = queue.faulty_blocks.getIndex(address) orelse return false;
+        const fault = &queue.faulty_blocks.entries.items(.value)[fault_index];
+        return fault.checksum == checksum and fault.state == .waiting;
+    }
+
+    pub fn repair_commence(queue: *const GridRepairQueue, address: u64, checksum: u128) void {
+        assert(queue.repair_ready(address, checksum));
+
+        const fault_index = queue.faulty_blocks.getIndex(address).?;
+        const fault = &queue.faulty_blocks.entries.items(.value)[fault_index];
+        assert(fault.checksum == checksum);
+        assert(fault.state == .waiting);
+
+        if (fault.progress == .table_content) {
+            const progress = &fault.progress.table_content;
+            assert(progress.table.table_blocks_written < progress.table.table_blocks_total.?);
+            assert(!progress.table.content_blocks_received.isSet(progress.index));
+
+            progress.table.content_blocks_received.set(progress.index);
+        }
+
+        fault.state = .writing;
+    }
+
+    pub fn repair_complete(queue: *GridRepairQueue, block: BlockPtrConst) void {
+        const block_header = schema.header_from_block(block);
+        const fault_index = queue.faulty_blocks.getIndex(block_header.op).?;
+        const fault_address = queue.faulty_blocks.entries.items(.key)[fault_index];
+        const fault: FaultyBlock = queue.faulty_blocks.entries.items(.value)[fault_index];
+        assert(!queue.canceling);
+        assert(fault_address == block_header.op);
+        assert(fault.checksum == block_header.checksum);
+        assert(fault.state == .aborting or fault.state == .writing);
+
+        queue.release_fault(fault_index);
+
+        if (fault.state == .aborting) {
+            queue.checkpointing.?.aborting -= 1;
+            return;
+        }
+
+        if (fault.progress == .table_index) {
+            // The reason that the content blocks are queued here (when the write ends) rather
+            // than when the write begins is so that a `request_block()` can be converted to a
+            // `request_table()` after the former's write is already in progress.
+            queue.request_table_content(fault.progress.table_index.table, block);
+        }
+
+        if (switch (fault.progress) {
+            .block => null,
+            .table_index => |progress| progress.table,
+            .table_content => |progress| progress.table,
+        }) |table| {
+            assert(table.table_blocks_total != null); // We already received the index block.
+            assert(table.table_blocks_written < table.table_blocks_total.?);
+            assert(table.content_blocks_received.count() <= table.table_blocks_total.? - 1);
```

Comment:
If this could be done without duplciating this bit, assert that `content_blocks_received.isSet(fault.progerss.index)` on the `table_content` branch, and that it is zero on `table_index` branhc.